lib/rust/metamodel/src/java/implementation.rs

```diff
@@ -0,0 +1,253 @@
+//! Generation of Java syntax implementing datatypes.
+
+use crate::java::*;
+use std::fmt::Write;
+
+
+
+// ===================================
+// === Implementing Java Datatypes ===
+// ===================================
+
+/// Implement a system of datatypes.
+pub fn implement(graph: &TypeGraph, package: &str) -> Vec<syntax::Class> {
+    let mut implementations = BTreeMap::new();
+    for (id, class) in graph.classes.iter() {
+        if !class.builtin {
+            implementations.insert(id, implement_class(graph, id));
+        }
+    }
+    for (id, class) in graph.classes.iter() {
+        if let Some(parent) = class.parent {
+            let mut inner = implementations.remove(&id).unwrap();
+            inner.static_ = true;
+            implementations.get_mut(&parent).unwrap().nested.push(inner);
+        }
+    }
+    for class in implementations.values_mut() {
+        class.package = Some(package.to_owned());
+    }
+    implementations.into_values().collect()
+}
+
+/// Get a type's qualified name, relative to the generated package.
+pub fn path(graph: &TypeGraph, id: ClassId) -> String {
+    let mut components = vec![];
+    let mut next_id = Some(id);
+    while let Some(id) = next_id {
+        let ty = &graph[id];
+        components.push(ty.name.as_str());
+        next_id = ty.parent;
+    }
+    components.reverse();
+    components.join(".")
+}
+
+fn class_fields_<'v, 's: 'v, 'c: 'v>(
+    graph: &'s TypeGraph,
+    class: &'c Class,
+    out: &mut Vec<&'v Field>,
+    start: Option<usize>,
+    end: Option<usize>,
+) {
+    let mut fields = &class.fields[..];
+    if let Some(end) = end {
+        fields = &fields[..end];
+    }
+    if let Some(start) = start {
+        fields = &fields[start..];
+    } else if let Some(parent) = class.parent {
+        let index = Some(graph[parent].child_field.unwrap());
+        class_fields_(graph, &graph[parent], out, None, index);
+        out.extend(fields);
+        class_fields_(graph, &graph[parent], out, index, None);
+        return;
+    }
+    out.extend(fields);
+}
+
+/// Get the fields owned by a class, including its own fields and the fields of its supertypes.
+pub fn class_fields<'v, 's: 'v, 'c: 'v>(graph: &'s TypeGraph, class: &'c Class) -> Vec<&'v Field> {
+    let mut out = vec![];
+    class_fields_(graph, class, &mut out, None, None);
+    out
+}
+
+/// Produce syntax referring to the given type.
+pub fn quote_type(graph: &TypeGraph, data: &FieldData) -> syntax::Type {
+    let class = match data {
+        FieldData::Object { type_, .. } => return quote_class_type(graph, *type_),
+        FieldData::Primitive(Primitive::Int { .. }) => "int",
+        FieldData::Primitive(Primitive::Bool) => "boolean",
+        FieldData::Primitive(Primitive::Long { .. }) => "long",
+    };
+    syntax::Type::named(class)
+}
+
+/// Produce syntax referring to the given class.
```

Comment:
Extend the docs, please. It's not obvious what "syntax referring to the given class" means.